### PR TITLE
warp: validate request host header field

### PR DESCRIPTION
**Pull request by sir-murray**
See #503

### CODE REVIEW

File: warp/Network/Wai/Handler/Warp/RequestHeader.hs [hunk 1]
```diff
@@ -35,8 +35,10 @@ parseHeaderLines [] = throwIO $ NotEnoughLines []
 parseHeaderLines (firstLine:otherLines) = do
     (method, path', query, httpversion) <- parseRequestLine firstLine
     let path = H.extractPath path'
-        hdr = map parseHeader otherLines
-    return (method, path', path, query, httpversion, hdr)
+        (hdr, hostCount) = foldr parseValidHeader ([], 0) otherLines
+    if hostCount == 1
```
Comment (meteficha) [inline warp/Network/Wai/Handler/Warp/RequestHeader.hs]: Now that the result of fold is being used strictly, perhaps it should be changed into a `foldl'`?

**Comment by meteficha** [comment]
Very nice! :+1: :shipit:

File: warp/Network/Wai/Handler/Warp/RequestHeader.hs (hunk 1 above)
Comment (sir-murray) [inline warp/Network/Wai/Handler/Warp/RequestHeader.hs]: I don't know, I'm fairly new to Haskell so if you say so I'll change it.

**Comment by sir-murray** [comment]
@meteficha Changed it to foldl'

**Comment by kazu-yamamoto** [comment]
We need to be very careful about this kind of change. This patch introduce one more traversal of headers with the very heavy operation (i.e. (==)). Moreover, this patch does not cover HTTP/2.

- First, we should continue to discuss whether or not we implement this feature in #503.
- If we agree, we need to find the best place for this traversal. I guess `traverseHeader` is a candidate for HTTP/1.1.
- Confirm that this change does not introduce performance regression.

**Comment by sir-murray** [comment]
ok, will close this for now.

**Pull request closed without merging**